NullDates are different across databases

Bakual opened the pull request:
### Issue

Since the various databases have different null dates the check in the JHtml::calendar failes on non-MySQL databases. The result is that if you edit for example an article and save it again, the "Finish Publishing" date will be populated with "1970-01-01" and the article is expired and no longer shows.
See https://groups.google.com/d/msg/joomla-dev-cms/R8iw4TDgbZo/MmTUcIfoOMEJ for the original issue message.
### Testing

Test if the date fields act as expected both when there is a value saved and if left empty.

Commit:
NullDates are different across databases

Commit:
Use the database nulldate in frontend for check

Commit:
Adjusting NullDate to include the miliseconds.

Commit:
Only convert the value when not a nullDate

Commit:
Don't show tooltip

brianteeman commented:
Where are we with this?
Without this being committed for 3.4 once again we will be making a release that can not work with all the supported databases

brianteeman commented:
Raising priority according to https://docs.joomla.org/Bug_and_Issue_Tracker_Priority<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3803">issues.joomla.org/joomla-cms/3803</a>.</sub>

Bakual commented:
Imho it should work as is. I have no clue where the issues from @waader are coming from. It may well be unrelated issues which just surfaced because of this PR.

wilsonge commented:
Two different people reported the mssql issues. We need someone with it who can look into the issue for us before we can merge this imho :(

brianteeman commented:
We have two harsh options
1. Commit this now so that at least postgres is working in 3.4 
2. Ignore this now - postgres will be broken again in 3.4 but mssql works

Personally I would go with option 1. If the MS SQL users dont help testing etc...

Bakual commented:
The only change I can imagine which is related to MS SQL and storing articles would be the added miliseconds to the getNullDate method of MS SQL. I could remove that for now and then it should still work for PostgreSQL (since nothing changes for that) and MS SQL should work again as bad as it does today. There should be no SQL errors anymore, at least I can't imagine where they would come from.
Would that be an approach?
After that one could make a PR to solve the MS SQL issue with the added miliseconds.

brianteeman commented:
Go for it or we will be stuck forever (or until PLT get some windows test servers)

Commit:
Adjusting NullDate to include the miliseconds. (reverted from commit ccd4d3afa2f059f27c40c94aedfae8375a620ffc)

Bakual commented:
Reverted the commit which added the miliseconds to the MS SQL getNullDate method.
Hopefully it works better now.

Also resolved a conflict and rebased with staging.

roland-d commented:
This PR has been committed in commit b4a0cde but I forgot to add the # to the PR number.

roland-d commented:
Commit 65ee68b fixes #3803

Bakual commented:
@roland-d It will never say "merged" if you close it from a commit. That's because you didn't merge it :smile:
It will instead say `@roland-d closed this pull request from a commit x hours ago` and `roland-d closed this in abcdefg x hours ago`
The PR will show as `Closed with unmerged commits` for the PR owner.

That is fine and expected since you took the changes, squashed them and commited them yourself.

sweco-semhul commented:
The issue for MS SQL do to milliseconds(.000) never got solved?

Bakual commented:
It should be. Since Joomla 3.4.0.

sweco-semhul commented:
From what I can see its not. https://github.com/joomla/joomla-cms/blob/3.4.0/libraries/joomla/database/driver/sqlsrv.php#L46
and not in master either https://github.com/joomla/joomla-cms/blob/master/libraries/joomla/database/driver/sqlsrv.php#L46

Or am I missing something?

Bakual commented:
Ah you're right. I had to revert that part due to some issues I couldn't solve. Didn't remember that.
Feel free to attempt to fix it. I can't test it myself so I could only shoot into the dark and hope for the best.

sweco-semhul commented:
Sure, I have done the change in our test environment and it seems to run whit out any problems. Do you remember what the issue where?

wilsonge commented:
https://github.com/joomla/joomla-cms/pull/3803#issuecomment-59619453 I'd guess it's the one in this comment
